Adding accessibility role and state to bottom bar

fawcilize opened the pull request:
### Motivation

Exposing `accessibilityRole` and `accessibilityStates` through props in the same way as `accessibilityLabel`. This allows screen readers to announce things like: `Selected. Banana. Button.`

Reasonable defaults were provided, however, consumers can provide custom overrides in the cases where the role might be `imagebutton` or the state might be `disabled`.

### Test plan

Select a button with VoiceOver (iOS) or Talkback (Android) enabled. The screen reader should read something like `Selected. <AccessibilityLabel>. Button.` or something slightly different depending on which platform you are on. Buttons that are not focused should not have `Selected` announced.

fawcilize commented on src/views/BottomTabBar.js:
```diff
@@ -240,6 +251,9 @@ class TabBarBottom extends React.Component<Props> {
               accessibilityLabel={accessibilityLabel}
+              accessibilityRole={accessibilityRole}
+              accessibilityStates={accessibilityStates}
+              focused={focused}
```
I also exposed this to the `ButtonComponent` to handle the case of a custom component. When using a custom component, I was not able to find a way to determine whether or not that it was focused, so I could not determine `accessibilityStates`.

satya164 replied:
Let's keep this PR focused on accessibility. I think there is some way to check if it's focused and I'm not sure about the API. So I'd have to spend more time to merge it if it's going to be in the same PR

fawcilize replied:
I removed the focused property for this PR. Though, it's still unclear how a custom component would become aware of whether or not it's focused without it. However, I think with the other changes, I would no longer need to create a custom component and could just use the default.

satya164 commented on src/views/BottomTabBar.js:
```diff
@@ -220,9 +226,14 @@ class TabBarBottom extends React.Component<Props> {
-            route,
-          });
+          const accessibilityLabel = this.props.getAccessibilityLabel({ route });
+          const accessibilityRole = this.props.getAccessibilityRole({ route }) || 'button';
```
What other accessibility roles instead of button do you have in mind?

fawcilize replied:
I think having `button` is a good default, but if someone only had an icon with no label, they may want to use `imagebutton`. Overriding `getAccessibilityRole` would allow that.

fawcilize replied:
The full list of available roles are here: https://facebook.github.io/react-native/docs/accessibility#accessibilityrole-ios-android

satya164 commented on src/views/BottomTabBar.js:
```diff
@@ -220,9 +226,14 @@ class TabBarBottom extends React.Component<Props> {
+
+          let accessibilityStates = this.props.getAccessibilityStates({ route });
+          if (!accessibilityStates) {
+            accessibilityStates = focused ? ['selected'] : [];
```
What other accessibility states instead of selected/not selected do you have in mind?

fawcilize replied:
`selected` and `disabled` are the only options allowed through the API, but this library would not really know if a tab is disabled (perhaps for an app's custom permissions?). If someone had a disabled tab, they could override `getAccessibilityStates` and return `['disabled']` or even `['selected', 'disabled']`.

orzarchi commented:
Any chance of this PR getting some attention? This is currently the only hold out on making my entire app accessible.

fawcilize commented:
There has been a merge conflict since I first opened the PR, so I pushed up a fix.

@satya164 Was there something else you'd like me to address?

satya164 commented:
Sorry for the delay. Can you rebase against master so the CI runs? I'll merge it after the CI passes.

fawcilize commented:
I rebased and fixed some linting errors. I did not include the states on the MaterialTopBar because they are not exposed by the underlying lib `react-native-tab-view`

MoOx commented:
Any reason 'tab' didn't end up being the default? I tried some apple native apps, and they all say "<label>. Tab. <x> of <y>" (I guess the parent need to be set with `accessibilityRole="tablist"`.

satya164 commented:
> Any reason 'tab' didn't end up being the default? I tried some apple native apps, and they all say ". Tab. of "

I remember that's what we used to do. Maybe accidentally changed it.

> I guess the parent need to be set with accessibilityRole="tablist"

AFAIK RN doesn't expose this value?

If you could send a PR to fix it, it'll be awesome.

MoOx commented:
<img width="740" alt="image" src="https://user-images.githubusercontent.com/157534/62203494-97326880-b38b-11e9-93f4-31524e210887.png">

Ok I will see what I can do :)

satya164 commented:
Ah, must be a recent addition 😅
